### ISSUE TITLE
group positioning not working with YAML inventory

### ISSUE THREAD

**Issue by michaelkebe** [ansible/ansible#69908]
##### SUMMARY

When using an inventory in YAML format the positioning does not work properly.

The problem is that YAML does not have an ordering of keys (which is used in the inventory YAML file). This is an implementation detail of the YAML parser. The currently presented ordering is undefined (not speficied in the YAML spec).

My suggestions:
- Report an error (a warning at least), when using group positioning and YAML inventory files.
- Allow only the use of lists for the hosts key in the YAML inventory (NOT backward-compatible)
- Make a huge note in the documentation (https://docs.ansible.com/ansible/latest/user_guide/intro_patterns.html#using-group-position-in-patterns)

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

Patterns: targeting hosts and groups.
https://docs.ansible.com/ansible/latest/user_guide/intro_patterns.html#using-group-position-in-patterns

##### ANSIBLE VERSION

```paste below
ansible 2.9.9
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/XXXXXXX/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.17 (default, Apr 15 2020, 17:20:14) [GCC 7.5.0]
```

##### CONFIGURATION

```paste below

```

##### OS / ENVIRONMENT

Ubuntu 18.04.3 LTS

##### STEPS TO REPRODUCE

Clone this gist https://gist.github.com/michaelkebe/cd714b71dd3bb22d2122b162ea0973e7 and run run_ini.sh and run_yml.sh and compare the outputs.

hosts.ini:
```yaml
[my_group]
host-1
host-2
host-3
```
hosts.yml:
```yaml
all:
  children:
    my_group:
      hosts:
        host-1:
        host-2:
        host-3:
```
playbook.yml:
```yaml
---
- hosts: my_group
- hosts: my_group[0]
- hosts: my_group[1]
- hosts: my_group[2]
```

##### EXPECTED RESULTS

The outputs of the two script should be the same (output of run_ini.sh). It should be taken account of the ordering of the hosts in the YML inventory.

##### ACTUAL RESULTS

The host selection of the YAML inventory is wrong. E.g. my_group[0] should return host-1, but returns host-3. The output when using the INI inventory is correct.

```paste below
$ sh run_ini.sh
@all:
  |--@my_group:
  |  |--host-1
  |  |--host-2
  |  |--host-3
  |--@ungrouped:

playbook: playbook.yml

  play #1 (my_group): my_group  TAGS: []
    pattern: [u'my_group']
    hosts (3):
      host-3
      host-2
      host-1

  play #2 (my_group[0]): my_group[0]    TAGS: []
    pattern: [u'my_group[0]']
    hosts (1):
      host-1

  play #3 (my_group[1]): my_group[1]    TAGS: []
    pattern: [u'my_group[1]']
    hosts (1):
      host-2

  play #4 (my_group[2]): my_group[2]    TAGS: []
    pattern: [u'my_group[2]']
    hosts (1):
      host-3

$ sh run_yml.sh
@all:
  |--@my_group:
  |  |--host-1
  |  |--host-2
  |  |--host-3
  |--@ungrouped:

playbook: playbook.yml

  play #1 (my_group): my_group  TAGS: []
    pattern: [u'my_group']
    hosts (3):
      host-3
      host-2
      host-1

  play #2 (my_group[0]): my_group[0]    TAGS: []
    pattern: [u'my_group[0]']
    hosts (1):
      host-3

  play #3 (my_group[1]): my_group[1]    TAGS: []
    pattern: [u'my_group[1]']
    hosts (1):
      host-2

  play #4 (my_group[2]): my_group[2]    TAGS: []
    pattern: [u'my_group[2]']
    hosts (1):
      host-1

```

**Comment by sivel** [comment]
Duplicate of the recently resolved https://github.com/ansible/ansible/issues/34861

**Comment by michaelkebe** [comment]
I don't agree with the close.

In my opinion ansible should only allow YAML sequences to preserve the
order and should raise an error, if group positioning is used with an YAML
inventory with hostnames as keys.

I know, it is a breaking chance, but the YAML spec is clear:

"In every case where node order is significant, a sequence must be used."

https://yaml.org/spec/1.2/spec.html#id2765608

Matt Martz <notifications@github.com> schrieb am Fr., 5. Juni 2020, 16:53:

**Comment by bcoca** [comment]
@michaelkebe please read the message, this has recently been resolved via #58000 which does what you ask for.
